File: First Missing Positive/First Missing Positive.cpp
class Solution {
public:
    int firstMissingPositive(int A[], int n) {
        for(int i = 0; i < n; ){
            if(A[i] != i + 1 && A[i] > 0 && A[i] <= n && A[A[i] - 1] != A[i]){
                int targetIndex = A[i] - 1;
                int temp = A[i];
                A[i] = A[targetIndex];
                A[targetIndex] = temp;
            }
            else{
                ++i;
            }
        }
        for(int i = 0; i < n; ++i){
            if(A[i] != i + 1)       return i + 1;
        }
        return n + 1;
    }
};

File: First Missing Positive/First Missing Positive.java
public class Solution {
    public int firstMissingPositive(int[] A) {
        for(int i = 0; i < A.length;){
            if(A[i] > 0 && A[i] <= A.length && A[i] != i + 1 && A[i] != A[A[i] - 1]){
                int temp = A[i];
                A[i] = A[A[i] - 1];
                A[temp - 1] = temp;
            }
            else{
                ++i;
            }
        }
        for(int i = 0; i < A.length; ++i){
            if(A[i] != i + 1){
                return i + 1;
            }
        }
        return (A.length == 0) ? 1 : A.length + 1;
    }
}
